File: 5b.pl
#! /usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use ARGV::OrDATA;

my @grid;
my $danger_tally = 0;
while (<>) {
    my ($x1, $y1, $x2, $y2) = /(.+),(.+) -> (.+),(.+)/;
    my $xstep = $x2 <=> $x1;
    my $ystep = $y2 <=> $y1;

    my ($x, $y) = ($x1, $y1);
    while (1) {
        ++$danger_tally if 1 == $grid[$x][$y]++;
        last if $x == $x2 && $y == $y2;

        $x += $xstep;
        $y += $ystep;
    }
}
say $danger_tally;

__DATA__
0,9 -> 5,9
8,0 -> 0,8
9,4 -> 3,4
2,2 -> 2,1
7,0 -> 7,4
6,4 -> 2,0
0,9 -> 2,9
3,4 -> 1,4
0,0 -> 8,8
5,5 -> 8,2

File: 5b.py
#! /usr/bin/python3
import sys

grid = {}
danger_tally = 0

with open(sys.argv[1], 'r') as fh:
    for line in fh:
        ((x1, y1), (x2, y2)) = map(lambda coord: map(int, coord.split(',')),
                                   line.split(' -> '))

        xstep = 1 if x1 < x2 else 0 if x1 == x2 else -1
        ystep = 1 if y1 < y2 else 0 if y1 == y2 else -1
        (x, y) = (x1, y1)

        while True:
            if not x in grid:
                grid[x] = {}
            if not y in grid[x]:
                grid[x][y] = 0
            grid[x][y] += 1
            if grid[x][y] == 2:
                danger_tally += 1
            if x == x2 and y == y2:
                break
            x += xstep
            y += ystep

print(danger_tally)
